src/hdmf/common/resources.py

```diff
@@ -0,0 +1,355 @@
+        used within the same object and field. To do so, this method must be called for the two different resources.
+        The returned Key objects must be managed by the caller so as to be appropriately passed to subsequent calls
+        to methods for storing information about the different resources.
+        """
+        return Key(key, table=self.keys)
+
+    @docval({'name': 'key', 'type': (str, Key), 'doc': 'the key to associate the resource with'},
+            {'name': 'resource_name', 'type': str, 'doc': 'the name of the resource'},
+            {'name': 'resource_entity_id', 'type': str, 'doc': 'the entity at the resource to associate'},
+            {'name': 'resource_entity_uri', 'type': str, 'doc': 'the URL for the entity at the resource'})
+    def add_resource(self, **kwargs):
+        """
+        Add an external resource that will be referenced to using the given key
+        """
+        key = kwargs['key']
+        resource_name = kwargs['resource_name']
+        resource_entity_id = kwargs['resource_entity_id']
+        resource_entity_uri = kwargs['resource_entity_uri']
+        if not isinstance(key, Key):
+            key = self.add_key(key)
+        resource_entity = ResourceEntity(key, resource_name, resource_entity_id, resource_entity_uri,
+                                         table=self.resources)
+        return resource_entity
+
+    @docval({'name': 'container', 'type': (Container, str), 'doc': 'the Container to add'},
+            {'name': 'field', 'type': str, 'doc': 'the field on the Container to add'})
+    def add_object(self, **kwargs):
+        """
+        Add an object that references an external resource
+        """
+        container, field = popargs('container', 'field', kwargs)
+        if isinstance(container, Container):
+            container = container.object_id
+        obj = Object(container, field, table=self.objects)
+        return obj
+
+    @docval({'name': 'obj', 'type': (int, Object), 'doc': 'the Object to that uses the Key'},
+            {'name': 'key', 'type': (int, Key), 'doc': 'the Key that the Object uses'})
+    def add_external_reference(self, **kwargs):
+        """
+        Specify that an object (i.e. container and field) uses a key to reference
+        an external resource
+        """
+        obj, key = popargs('obj', 'key', kwargs)
+        return ObjectKey(obj, key, table=self.object_keys)
+
+    def _check_object_field(self, container, field):
+        """
+        A helper function for checking if a container and field have been added.
+
+        If the container and field have not been added, add the pair and return
+        the corresponding Object. Otherwise, just return the Object.
+        """
+        object_id = self.objects.which(object_id=container)
+        if len(object_id) > 0:
+            field_id = self.objects.which(field=field)
+            object_id = list(set(object_id) & set(field_id))
+
+        if len(object_id) == 1:
```

Comment:
I suggest adding a check here if len(object_id) > 1: raise warning and return the first one. We cannot guarantee that other APIs are ensuring only one row for a given (object, field)